File: swift/solver.swift
import Foundation

func getRow(_ puzzle: [Int?], row: Int) -> [Int?] {
    return puzzle.enumerated().filter({ 
        (i, element) in i >= row * 9 && i < (row + 1) * 9
    }).map({
        (i, element) in element
    })
}

func getColumn(_ puzzle: [Int?], col: Int) -> [Int?] {
    return puzzle.enumerated().filter({ 
        (i, element) in i % 9 == col 
    }).map({
        (i, element) in element
    })
}

func getSquareIndex(_ index: Int) -> Int {
    return (index % 9) / 3 + ((index / 9) / 3) * 3
}

func getSquare(_ puzzle: [Int?], square: Int) -> [Int?] {
    return puzzle.enumerated().filter({ 
        (i, element) in getSquareIndex(i) == square 
    }).map({
        (i, element) in element
    })
}

func getPossibilities(row: [Int?], col: [Int?], square: [Int?]) -> Set<Int> {
    let used = Set((row + col + square).filter({ element in element != nil }).map({
        (element: Int?) -> Int in element!
    }))
    var possibilities = Set(1...9)
    possibilities.subtract(used)
    return possibilities
}

func solve(_ puzzle: [Int?]) -> [Int?] {
    var solved = Array(puzzle)
    var changed = true
    while changed {
        changed = false
        for i in 0..<solved.count {
            if solved[i] == nil {
                let possibilities = getPossibilities(
                    row: getRow(solved, row:i), 
                    col: getColumn(solved, col:i), 
                    square: getSquare(solved, square:getSquareIndex(i)))
                if let value = possibilities.first {
                    solved[i] = value
                    changed = true
                }
            }
        }
    }
    return solved
}

func draw(_ puzzle: [Int?]) {
    for i in 0..<puzzle.count {
        print(puzzle[i] ?? ".", terminator:"")
        if (i+1) % 9 == 0 {
            print("")
        }
    }
}

if (CommandLine.arguments.count < 2) {
    print("Usage: swift solver.swift <puzzle>")
    print("Use dots for empty spaces, e.g. 12..5.78.")
} else {
    let puzzle = CommandLine.arguments[1].characters.map({
        (square: Character) -> Int? in Int(String(square))
    })
    if puzzle.count != 81 {
        print("Puzzle is not a 9x9 Sudoku")
    }
    else {
        let solved = solve(puzzle)
        draw(solved)
    }
}

File: golang/solver/solver.go
package main

import "errors"
import "fmt"
import "os"

func main() {
	if len(os.Args) < 2 {
		fmt.Println("Usage: solver <puzzle>")
		fmt.Println("Use dots for empty spaces, e.g. 12..5.78.")
		return
	}

	puzzle, err := parsePuzzle(os.Args[1])
	if err != nil {
		fmt.Println("Error")
		fmt.Printf("Arg: %s, len: %d", os.Args[1], len(os.Args[1]))
		return
	}

	changed := true
	for changed {
		changed = false
		for i, v := range puzzle {
			if v == -1 {
				possibilities := possibilities(row(puzzle, i), column(puzzle, i), square(puzzle, i))
				if len(possibilities) == 1 {
					puzzle[i] = possibilities[0]
					changed = true
				}
			}
		}
	}

	fmt.Println(puzzle)
}

func parsePuzzle(input string) ([]int, error) {

	if len(input) != 81 {
		return nil, errors.New("Input not the correct size")
	}

	output := make([]int, 81)
	for i, v := range input {
		if v == '.' {
			output[i] = -1
		} else {
			output[i] = int(v - '0')
			if output[i] < 1 || output[i] > 9 {
				return nil, errors.New("Invalid character found in input")
			}
		}
	}
	return output, nil
}

func row(puzzle []int, index int) ([]int) {
	rowStart := index - (index % 9)
	var row []int
	return append(row, puzzle[rowStart:rowStart+9]...)
}

func column(puzzle []int, index int) ([]int) {
	var column []int
	for i := index%9; i < 81; i += 9 {
		column = append(column, puzzle[i])
	}
	return column
}

func square(puzzle []int, index int) ([]int) {
	squareNumber := (index % 9) / 3 + ((index / 9) / 3) * 3
	start := int(squareNumber / 3) * 27 + (squareNumber % 3) * 3
	
	var square []int
	square = append(square, puzzle[start:start+3]...)
	square = append(square, puzzle[start+9:start+9+3]...)
	square = append(square, puzzle[start+18:start+18+3]...)
	return square
}

func possibilities(row []int, column []int, square []int) ([]int) {
	s := map[int]bool {1: true, 2: true, 3: true, 4: true, 5: true, 6: true, 7: true, 8: true, 9: true}
	for _, num := range append(row, append(column, square...)...) {
		_, ok := s[num]
		if ok {
			s[num] = false
		}
	}
	p := make([]int, 0)
	for k, v := range s {
		if v {
			p = append(p, k)
		}
	}
	return p
}